src/EditorFeatures/Text/Extensions.TextBufferContainer.cs

```diff
@@ -37,8 +37,8 @@ private TextBufferContainer(ITextBuffer editorBuffer)
             /// <summary>
             /// A weak map of all Editor ITextBuffers and their associated SourceTextContainer
             /// </summary>
-            private static readonly ConditionalWeakTable<ITextBuffer, TextBufferContainer> s_textContainerMap = new ConditionalWeakTable<ITextBuffer, TextBufferContainer>();
-            private static readonly ConditionalWeakTable<ITextBuffer, TextBufferContainer>.CreateValueCallback s_createContainerCallback = CreateContainer;
+            private static readonly ConditionalWeakTable<ITextBuffer, WeakReference<TextBufferContainer>> s_textContainerMap = new();
+            private static readonly ConditionalWeakTable<ITextBuffer, WeakReference<TextBufferContainer>>.CreateValueCallback s_createContainerCallback = CreateContainer;
```

Comment:
❓ Why not just store a `WeakReference<TextBufferContainer>` in `ITextBuffer.Properties`?

Reply:
Generally, we could. As best I can recall there was much spiritous debate early in Roslyn about which pattern to use. The specific concern with using the text buffer property bag is it's public, so any malicious extension is able to get to your object without much work. And whereas somebody could of course use reflection to get at this, that's clearly an "out of bounds" operation. Going into the property bag is much less "out of bounds". For example, there's plenty of extensions that get the ITextDocument for an ITextBuffer by simply doing .GetProperty(typeof(ITextDocument)) and I think that was more or less the _supported_ way.

Reply:
Let's switch to using `Properties` with an opaque key (singleton object defined in Roslyn).

Reply:
@sharwell if we want to do that, let's take that to the design meeting and make that change consistently across the other ConditionalWeakTables -- we've got a bunch more of these.